### PR TITLE
chore: Deploy and Retrieve

### PR DESCRIPTION
**Pull request by CristiCanizales**
### What does this PR do?
Implementation for Deploy and Retrieve

### What issues does this PR fix or reference?
@W-13068481@

See also https://salesforce.quip.com/qVEkABuGd7ao and https://docs.google.com/spreadsheets/d/1DwUs_Xb-bGB-1XKZsRurN2l3O6Uxq5n68QizkWRNcSI/edit#gid=0

### How to test:

1. Go to the [End to End Tests Workflow](https://github.com/forcedotcom/salesforcedx-vscode/actions/workflows/e2e.yml) in salesforcedx-vscode repo
2. Paste the PR's branch (`cristi/deploy-and-retrieve`) in the second field of the Run workflow dropdown (**Set the branch to use for automation tests**)
3. Choose deployAndRetrieve.e2e.ts from the dropdown to select the test to run.

AC:
All steps within the testsuite should pass.

Successful runs: 
- [5616795350](https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/5616795350) ✅ 
- [5662691022](https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/5662691022) ✅

### CODE REVIEW

File: test/specs/deployAndRetrieve.e2e.ts [hunk 1]
```diff
@@ -5,26 +5,330 @@
...
+    await preferDeployOnSaveBtn.click();
+    await utilities.pause(1);
+
+    // Reload window to apply settings
+    await utilities.runCommandFromCommandPrompt(workbench, 'Developer: Reload Window', 10);
+
+    // Clear all notifications so clear output button is reachable
+    await utilities.runCommandFromCommandPrompt(
+      workbench,
+      'Notifications: Clear All Notifications',
```
Comment (CristiCanizales) [inline test/specs/deployAndRetrieve.e2e.ts:294]: This was preventing the 'clear output' button from being visible/reachable

File: test/specs/deployAndRetrieve.e2e.ts [hunk 2]
```diff
@@ -5,26 +5,331 @@
...
+      'Salesforce CLI',
+      'Starting SFDX: Deploy Source to Org',
+      10
+    );
+    utilities.log('Deploy time - 3: ' + (await utilities.getOperationTime(outputPanelText!)));
+    expect(outputPanelText).not.toBeUndefined();
+    expect(outputPanelText).toContain('Deployed Source');
+    expect(outputPanelText).toContain(`Changed  MyClass    ApexClass  ${pathToClass}.cls`);
+    expect(outputPanelText).toContain(`Changed  MyClass    ApexClass  ${pathToClass}.cls-meta.xml`);
+    expect(outputPanelText).toContain('ended SFDX: Deploy Source to Org');
```
Comment (RitamAgrawal) [inline test/specs/deployAndRetrieve.e2e.ts:177]: To compare the deploy duration, How about we repeat the above deploy steps with the source tracking setting disabled?

Comment (RitamAgrawal) [reply test/specs/deployAndRetrieve.e2e.ts:177]: And we need to turn on source tracking in the org setting.

File: test/utilities/apexUtils.ts [hunk 3]
```diff
@@ -9,7 +9,11 @@ import { TextEditor } from 'wdio-vscode-service';
 import { runCommandFromCommandPrompt } from './commandPrompt';
 import { pause } from './miscellaneous';
 
-export async function createApexClassWithTest(name: string): Promise<void> {
+export async function createApexClass(
+  name: string,
+  classText: string,
+  breakpoint?: number
+): Promise<void> {
   const workbench = await browser.getWorkbench();
```
Comment (RitamAgrawal) [inline test/utilities/apexUtils.ts:17]: I think we can avoid pause with `.wait()` on workbench

File: test/utilities/apexUtils.ts [hunk 4]
```diff
@@ -9,7 +9,11 @@ import { TextEditor } from 'wdio-vscode-service';
 import { runCommandFromCommandPrompt } from './commandPrompt';
 import { pause } from './miscellaneous';
 
-export async function createApexClassWithTest(name: string): Promise<void> {
+export async function createApexClass(
```
Comment (RitamAgrawal) [inline test/utilities/apexUtils.ts:12]: 🎉

File: test/utilities/outputView.ts [hunk 5]
```diff
@@ -103,3 +98,42 @@ export async function attemptToFindOutputPanelText(
 
   return undefined;
 }
+
+export async function getOperationTime(outputText: string): Promise<string> {
+  const initialTime = outputText.substring(
```
Comment (RitamAgrawal) [inline test/utilities/outputView.ts:103]: When are we considering the initial and end time of operation here?

**Review by RitamAgrawal — COMMENTED** [review]
Amazing work 🎉 Have few suggestion to make it more robust in terms of taking source tracking in account. And few questions on operation time calculation.

File: test/utilities/outputView.ts (hunk 5 above)
Comment (CristiCanizales) [reply test/utilities/outputView.ts:103]: it's in the output text

File: test/utilities/apexUtils.ts (hunk 3 above)
Comment (CristiCanizales) [reply test/utilities/apexUtils.ts:17]: yep, that's in my other pr

File: test/specs/apexLsp.e2e.ts [hunk 6]
```diff
@@ -39,8 +39,12 @@ describe('Apex LSP', async () => {
...
     utilities.log(`${testSetup.testSuiteSuffixName} - Go to Definition`);
     // Get open text editor
-    const workbench = await browser.getWorkbench();
-    const editorView = workbench.getEditorView();
+    const workbench = await (await browser.getWorkbench()).wait();
+    const inputBox = await utilities.runCommandFromCommandPrompt(workbench, 'Go to File...', 1);
+    await inputBox.setText('ExampleClassTest.cls');
+    await inputBox.confirm();
+    await utilities.pause(1);
+    const editorView = await workbench.getEditorView().wait();
```
[the remark this answers is not in the document]
Comment (CristiCanizales) [reply test/specs/apexLsp.e2e.ts:47]: they're actually not needed anymore. Thanks

**Comment by RitamAgrawal** [comment]
Run [#114](https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/5673469825/job/15374984282) failing on step Push Deploy on save for mac

File: test/specs/deployAndRetrieve.e2e.ts [hunk 7]
```diff
@@ -5,26 +5,331 @@
...
+      workbench,
+      'Running SFDX: Deploy Source to Org',
+      utilities.FIVE_MINUTES
+    );
+    // At this point there should be no conflicts since this is a new class.
+    const successNotificationWasFound = await utilities.notificationIsPresent(
+      workbench,
+      'SFDX: Deploy Source to Org successfully ran'
+    );
+    expect(successNotificationWasFound).toBe(true);
```
Comment (RitamAgrawal) [inline test/specs/deployAndRetrieve.e2e.ts:319]: @CristiCanizales Is mac runner failing/ flaky on this expect statement?

Comment (CristiCanizales) [reply test/specs/deployAndRetrieve.e2e.ts:364]: no! The flaky part on remote is when trying to get the setting

File: test/utilities/outputView.ts (hunk 5 above)
Comment (RitamAgrawal) [reply test/utilities/outputView.ts:103]: So we are using the cli timestamp to account for operation time.

File: test/specs/deployAndRetrieve.e2e.ts [hunk 8]
```diff
@@ -5,26 +5,536 @@
...
+      1
+    );
+
+    const enableSourceTrackingBtn = await $(
+      'div[title="salesforcedx-vscode-core.experimental.enableSourceTrackingForDeployAndRetrieve"]'
+    );
+    await enableSourceTrackingBtn.click();
+    await utilities.pause(1);
+    // Reload window to update cache and get the setting behavior to work
+    await utilities.runCommandFromCommandPrompt(workbench, 'Developer: Reload Window', 10);
```
Comment (RitamAgrawal) [inline test/specs/deployAndRetrieve.e2e.ts:425]: Good capturing it in comments! A very easy to forget step!

**Comment by CristiCanizales** [comment]
> Run [#114](https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/5673469825/job/15374984282) failing on step Push Deploy on save for mac

I'll take care of this (running on remote) in another PR

**Review by RitamAgrawal — APPROVED** [review]

**Pull request merged**